From: Simon McVittie <smcv@collabora.com>
Date: Mon, 8 Feb 2021 10:28:25 +0000
Subject: io-channel test: Add coverage for g_io_channel_set_line_term(., .,
 -1)

Bug: https://gitlab.gnome.org/GNOME/glib/-/issues/2323
Forwarded: https://gitlab.gnome.org/GNOME/glib/-/merge_requests/1931
Signed-off-by: Simon McVittie <smcv@collabora.com>
---
 glib/tests/io-channel.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/glib/tests/io-channel.c b/glib/tests/io-channel.c
index ff53fce..4a1b108 100644
--- a/glib/tests/io-channel.c
+++ b/glib/tests/io-channel.c
@@ -49,8 +49,10 @@ test_read_line_embedded_nuls (void)
   channel = g_io_channel_new_file (filename, "r", &local_error);
   g_assert_no_error (local_error);
 
-  /* Only break on newline characters, not nuls. */
-  g_io_channel_set_line_term (channel, "\n", 1);
+  /* Only break on newline characters, not nuls.
+   * Use length -1 here to exercise glib#2323; the case where length > 0
+   * is covered in glib/tests/protocol.c. */
+  g_io_channel_set_line_term (channel, "\n", -1);
   g_io_channel_set_encoding (channel, NULL, &local_error);
   g_assert_no_error (local_error);
 
